tests/test_transfer.py

```diff
@@ -247,6 +249,98 @@ def test_disk_vs_tape_with_custom_strategy(rse_factory, root_account, mock_scope
+    _add_mock_queued_request(short_wait_rse_id, 10 * bytes_per_sec)
+    _add_mock_queued_request(long_wait_rse_id, 50 * bytes_per_sec)
+    _add_mock_queued_request(longest_wait_rse_id, 15 * bytes_per_sec)
+    _add_mock_queued_request(longest_wait_rse_id, 30 * bytes_per_sec)
+    _add_mock_queued_request(longest_wait_rse_id, 45 * bytes_per_sec)
+
+    # Below average wait sources should be picked ~90% of the time, above average wait sources ~10% of the time
+    # In their group, sources should be picked proportionally based on wait time (with shorter wait timers preferred)
+    above_average_exploration = .1
+    expected_prob_by_rse_name = {
+        shortest_wait_rse_name: (1 - above_average_exploration) * 2 / 3,
+        short_wait_rse_name: (1 - above_average_exploration) / 3,
+        long_wait_rse_name: above_average_exploration * 2 / 3,
+        longest_wait_rse_name: above_average_exploration / 3
+    }
+
+    # Since TransferWaitTime is randomized, we need to simulate lots of requests
+    using_transfer_wait_time = 'TransferWaitTime' in file_config_mock.get('transfers', 'source_ranking_strategies')
+    num_requests = 100 if using_transfer_wait_time else 1
+
+    files = [{'scope': mock_scope, 'name': 'lfn.' + generate_uuid(), 'type': 'FILE', 'bytes': 1, 'adler32': 'beefdead'} for _ in range(num_requests)]
+    dids = [{'scope': file['scope'], 'name': file['name']} for file in files]
+    for rse_id in [shortest_wait_rse_id, short_wait_rse_id, long_wait_rse_id, longest_wait_rse_id]:
+        add_replicas(rse_id=rse_id, files=files, account=root_account)
+
+    rule_core.add_rule(dids=dids, account=root_account, copies=1, rse_expression=dst_rse_name, grouping='ALL', weight=None, lifetime=None, locked=False, subscription_id=None)
+    topology = Topology().configure_multihop()
+    requests = list_and_mark_transfer_requests_and_source_replicas(rse_collection=topology, rses=all_rses)
+    [[_, transfers]] = pick_and_prepare_submission_path(topology=topology, protocol_factory=ProtocolFactory(),
+                                                        requests_with_sources=requests).items()
+
+    if using_transfer_wait_time:
+        picks_by_rse_name = defaultdict(lambda: 0)
+        for transfer in transfers:
+            picks_by_rse_name[transfer[0].src.rse.name] += 1
+
+        print("shortest_wait_rse_picks =", picks_by_rse_name[shortest_wait_rse_name])
+        print("short_wait_rse_picks =", picks_by_rse_name[short_wait_rse_name])
+        print("long_wait_rse_picks =", picks_by_rse_name[long_wait_rse_name])
+        print("longest_wait_rse_picks =", picks_by_rse_name[longest_wait_rse_name])
+
+        # We can model probability of picking a source using a binomial distribution.
+        # Standard deviation is sqrt(num_requests * .1 * .9). +- 3 standard deviation is 99.7% confidence interval
+        stdev = (num_requests * .1 * .9) ** 0.5
+        for rse_name, expected_prob in expected_prob_by_rse_name.items():
```

Comment:
99.7 % still means that 3 out of 1000 test runs can fail without there being any bug. This is a lot. considering that we run something like 20 tests in each run of the pipeline. Having tests randomly failing for no reason is not good. Even if the probability is low. Developers will start thinking they did something wrong in their code and can spend hours on a wild chase for an non-existing bug.

Reply:
I have changed it to +- 4 standard deviations. With 4 standard deviations, confidence interval is 99.9936% accurate (~1/15,625). I have also added a comment notifying folks that in the unlikely situation that it fails to simply re-run. Hopefully this is acceptable or is a randomized unit test always frowned upon?

For our test case, the st dev is 3 so +- 4 means +- 12 from the expected value. This is still meaningful to test because for example if expected value is 30, then confidence interval is [18, 42] and when expected value is 60, confidence interval is [48, 72] which do not overlap.